### DISCUSSION TITLE
Fixing Bazelisk Rolling Updates -- Stopped Correctly Detecting Updates

### DISCUSSION THREAD

**Discussion by cpsauer** [renovatebot/renovate#23766] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

Hello wonderful Renovate folks,

It is oh so very nice to have Renovate update the bazel version in `.bazelversion` files for Bazelisk, keeping folks using the latest while also keeping builds reproducible. (Feature launched last year--context in https://github.com/renovatebot/renovate/issues/6995#issuecomment-1161001802.)

But something seems to have gotten stuck, just for bazel rolling releases (~weekly releases, quite heavily tested but marked pre-release to distinguish from LTS).  Previously working auto-updates stopped--which means I know the approximate date updates stopped: on or after 6/9 but before the end of 7/12.

My attempt to scope: The update list is empty for .bazelversion files already pointing to a rolling release, but stable releases detect updates both to stable and rolling releases--but the wrong rolling release...a very old one rather than the newest. (`7.0.0-pre.20221026.2` rather than `7.0.0-pre.20230724.1`) I'm wondering if perhaps something happened in the sorting handling of their version formatting? What do you think could be going wrong?

Here's a minimal reproduction repo that shows the cases described above:
https://github.com/cpsauer/renovate-reproduction-bazelisk-not-updating
1.  No updates found for the rolling/.bazelversion file. (See logs.)
2. [The wrong update found for updating stable/.bazelversion to rolling.](https://github.com/cpsauer/renovate-reproduction-bazelisk-not-updating/pull/2)
3. [The right update found for keeping stable/.bazelversion on stable.](https://github.com/cpsauer/renovate-reproduction-bazelisk-not-updating/pull/1)

(Would love to transfer that repo over to renovate-reproductions if that's useful or a good home for it--please lmk how I should do so.)

Thanks so much for your time, patience, and all you do!
Chris

### Relevant debug logs

<details><summary>Logs</summary>

```
"bazelisk": [
      {
        "deps": [
          {
            "depName": "bazel",
            "currentValue": "7.0.0-pre.20230530.3",
            "datasource": "github-releases",
            "packageName": "bazelbuild/bazel",
            "updates": [],
            "versioning": "semver-coerced",
            "warnings": [],
            "sourceUrl": "https://github.com/bazelbuild/bazel",
            "registryUrl": "https://github.com",
            "currentVersion": "7.0.0-pre.20230530.3",
            "fixedVersion": "7.0.0-pre.20230530.3"
          }
        ],
        "packageFile": "rolling/.bazelversion"
      },
      {
        "deps": [
          {
            "depName": "bazel",
            "currentValue": "6.3.0",
            "datasource": "github-releases",
            "packageName": "bazelbuild/bazel",
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "6.3.2",
                "newValue": "6.3.2",
                "releaseTimestamp": "2023-08-08T17:10:42.000Z",
                "newMajor": 6,
                "newMinor": 3,
                "updateType": "patch",
                "branchName": "renovate/bazel-6.x"
              },
              {
                "bucket": "major",
                "newVersion": "7.0.0-pre.20221026.2",
                "newValue": "7.0.0-pre.20221026.2",
                "releaseTimestamp": "2022-11-02T18:54:19.000Z",
                "newMajor": 7,
                "newMinor": 0,
                "updateType": "major",
                "branchName": "renovate/bazel-7.x"
              }
            ],
            "versioning": "semver-coerced",
            "warnings": [],
            "sourceUrl": "https://github.com/bazelbuild/bazel",
            "registryUrl": "https://github.com",
            "currentVersion": "6.3.0",
            "isSingleVersion": true,
            "fixedVersion": "6.3.0"
          }
        ],
        "packageFile": "stable/.bazelversion"
      }
    ]
  }
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Comment by rarkins** [comment]
The "wrong" v7 update is because of https://github.com/cpsauer/renovate-reproduction-bazelisk-not-updating/blob/5e457c0453b3c693adb85fb84bbeb23cafa78319/renovate.json#L5C6-L5C28

**Comment by rarkins** [comment]
Fix: https://github.com/renovatebot/renovate/pull/23771

@viceice @secustor surprising to me to see semver-coerced causing a _problem_ here

**Comment by cpsauer** [comment]
Hey Rhys! Thanks for replying and taking action so fast--but the issue is something else, and persists with the release that includes that change, I'm pretty sure. (It reproduces the same in new runs, with the logs showing `"renovateVersion": "36.40.3"`, which is higher than the 36.40.2 in your PR.) I'm hoping I can ask for some more of your guy's help.

For clarity: The `"ignoreUnstable": false` is intentional and important; this is about the rolling bazel releases, which are  marked pre-release, as above. The expected (and former) behavior is that an update to `7.0.0-pre.20230724.1` would be proposed for both `rolling/.bazelversion` and `stable/.bazelversion` (in addition to the 6.3.2 update for the latter, already working). The proposed update to the much older `7.0.0-pre.20221026.2` should instead be the newer `7.0.0-pre.20230724.1`. I suspect that's the heart of the issue.

**Comment by rarkins** [comment]
Wanting pre release for a folder labeled "stable" is confusing and counter intuitive - was there a reason for doing so?

It would be better to reduce the reproduction to a single file/dependency and then demonstrate the problem again there

**Reply by cpsauer** [reply]
Yes. In my estimation, showing the two parallel cases significantly helps scope the issue by highlighting what might be going wrong--rolling version sorting, as above.

In any case, the main point is that that `rolling/.bazelversion` doesn't finds the available update to `7.0.0-pre.20230724.1` (noting the 2023 date embedded), but that updates to newer rolling versions used to work great.

**Reply by rarkins** [reply]
Please update this discussion once a newly minimized reproduction is available

**Reply by cpsauer** [reply]
I am now truly perplexed. I created the even more minimal https://github.com/cpsauer/renovate-reproduction-bazelisk-not-updating-just-rolling ... and things are now working there. But reruns of [the old reproduction](https://github.com/cpsauer/renovate-reproduction-bazelisk-not-updating) remain stuck with the same problems and with semver-coerced in the logs, both rerunning from the renovate logs site and with the retry PR checkbox. The original repo experiencing the problem is now unstuck and has received the correct update to .bazelversion--it is set to bypass PRs--but strangely it took a bit to get unstuck. Two runs with Renovate 36.40.3 noticed no update--but the following run with the same Renovate version, 36.40.3, found and applied the Bazel rolling release update. Any idea what's up?

**Reply by rarkins** [reply]
Could be result of the repo extract cache , or the package cache. But package cache is normally 15 minutes. Were there any commits to the base branch preceding it starting to work? Commits will invalidate/reset the extract cache

**Reply by cpsauer** [reply]
No commits pushed up in the time range where it started working. Good for me to learn that there's some cacheing and statefulness, though, for my future understanding.
(It looks like semver-coerced was also stuck and then changed to semver at the same time as things started working, if that's a good clue?)

**Reply by cpsauer** [reply]
Thanks so much for getting the core issue bazel rolling updating back working 🙏🏻 Really appreciate it.

**Reply by rarkins** [reply]
The intentions change to fix it was to use semver instead of coerced

**Reply by cpsauer** [reply]
Rhys, do you still want the reproduction repos for debugging the stuck cache issue? (If not, I'll spin them down in a week or so.) At the time of writing, the original reproduction does remain stuck on semver-coerced.

**Reply by rarkins** [reply]
I've pushed a small change just now which has the intention of invalidating any cached extracts for bazelisk. Hopefully that fixes the problem and we'll see it in the logs/PR

**Reply by cpsauer** [reply]
Seems to be unstuck, so I'll spin 'em down. Thanks again for fixing, Rhys!
